runtime/vm/ValueTypeHelpers.cpp

```diff
@@ -69,4 +70,115 @@ isClassRefQtype(J9Class *cpContextClass, U_16 cpIndex)
+                }
+        }
+        return index;
+}
+
+J9Class *
+findJ9ClassInFlattenedClassCache(J9FlattenedClassCache *flattenedClassCache, U_8 *className, UDATA classNameLength)
+{
+        /* first field indicates the number of classes in the cache */
+        UDATA length = flattenedClassCache->numberOfEntries;
+        J9Class *clazz = NULL;
+
+        for (UDATA i = 0; i < length; i++) {
+                J9UTF8* currentClassName = J9ROMCLASS_CLASSNAME(J9_VM_FCC_ENTRY_FROM_FCC(flattenedClassCache, i)->clazz->romClass);
+                if (J9UTF8_DATA_EQUALS(J9UTF8_DATA(currentClassName), J9UTF8_LENGTH(currentClassName), className, classNameLength)) {
+                        clazz = J9_VM_FCC_ENTRY_FROM_FCC(flattenedClassCache, i)->clazz;
+                        break;
+                }
+        }
+
+        Assert_VM_notNull(clazz);
+        return clazz;
+}
+
+UDATA
+getFlattenableFieldOffset(J9Class *fieldOwner, J9ROMFieldShape *field)
+{
+        Assert_VM_notNull(fieldOwner);
+        Assert_VM_notNull(field);
+
+        UDATA fieldOffset = 0;
+        J9FlattenedClassCache *flattenedClassCache = NULL;
+        J9ROMNameAndSignature *nameAndSig;
+        UDATA fieldIndex = 0;
+
+        nameAndSig = &field->nameAndSignature;
+        flattenedClassCache = fieldOwner->flattenedClassCache;
+        fieldIndex = findIndexInFlattenedClassCache(flattenedClassCache, nameAndSig);
+        Assert_VM_unequal(UDATA_MAX, fieldIndex);
+        J9FlattenedClassCacheEntry * flattenedClassCacheEntry = J9_VM_FCC_ENTRY_FROM_FCC(flattenedClassCache, fieldIndex);
+        fieldOffset = flattenedClassCacheEntry->offset;
+        return fieldOffset;
+}
+
+BOOLEAN
+isFlattenableFieldFlattened(J9Class *fieldOwner, J9ROMFieldShape *field)
+{
+        Assert_VM_notNull(fieldOwner);
+        Assert_VM_notNull(field);
+
+        BOOLEAN fieldFlattened = FALSE;
+        J9Class* clazz = getFlattenableFieldType(fieldOwner, field);
+        fieldFlattened = J9_ARE_ALL_BITS_SET((clazz)->classFlags, J9ClassIsFlattened);
+        return fieldFlattened;
+}
+
+J9Class *
+getFlattenableFieldType(J9Class *fieldOwner, J9ROMFieldShape *field)
+{
+        Assert_VM_notNull(fieldOwner);
+        Assert_VM_notNull(field);
+
+        J9Class * fieldType = NULL;
+        J9FlattenedClassCache *flattenedClassCache = NULL;
+        J9ROMNameAndSignature *nameAndSig;
+        UDATA fieldIndex = 0;
+
+        nameAndSig = &field->nameAndSignature;
+        flattenedClassCache = fieldOwner->flattenedClassCache;
+        fieldIndex = findIndexInFlattenedClassCache(flattenedClassCache, nameAndSig);
+        Assert_VM_unequal(UDATA_MAX, fieldIndex);
+        fieldType = J9_VM_FCC_ENTRY_FROM_FCC(flattenedClassCache, fieldIndex)->clazz;
+        return fieldType;
+}
+
+UDATA
+getFlattenableFieldSize(J9Class *fieldOwner, J9ROMFieldShape *field)
+{
+        Assert_VM_notNull(fieldOwner);
+        Assert_VM_notNull(field);
+
+        UDATA instanceSize = sizeof(fj9object_t);
```

Comment:
This is not legal - use J9VMTHREAD_REFERENCE_SIZE or J9JAVAVM_REFERENCE_SIZE (which means you'll need to pass in a vmThread or javaVM).